File: src/main/java/ca/rafemurray/ModItems.java
package ca.rafemurray;

import net.fabricmc.fabric.api.itemgroup.v1.FabricItemGroup;
import net.fabricmc.fabric.api.itemgroup.v1.ItemGroupEvents;
import net.minecraft.item.*;
import net.minecraft.registry.Registries;
import net.minecraft.registry.Registry;
import net.minecraft.registry.RegistryKey;
import net.minecraft.registry.RegistryKeys;
import net.minecraft.text.Text;
import net.minecraft.util.Identifier;
import java.util.function.Function;

public class ModItems {
    // Make and register a custom item group for the mod (creative inventory tab)
    public static final RegistryKey<ItemGroup> CUSTOM_ITEM_GROUP_KEY = RegistryKey.of(Registries.ITEM_GROUP.getKey(), Identifier.of(ElytraRacing.MOD_ID, "item_group"));
    public static final ItemGroup CUSTOM_ITEM_GROUP = FabricItemGroup.builder()
            .icon(() -> new ItemStack(Items.ELYTRA))
            .displayName(Text.translatable("itemGroup.elytra-racing"))
            .build();
    public static Item register(String name, Function<Item.Settings, Item> itemFactory, Item.Settings settings) {
        RegistryKey<Item> itemKey = RegistryKey.of(RegistryKeys.ITEM, Identifier.of(ElytraRacing.MOD_ID, name));
        Item item = itemFactory.apply(settings.registryKey(itemKey));
        Registry.register(Registries.ITEM, itemKey, item);
        return item;
    }

    public static final Item TEST = register("test", Item::new, new Item.Settings());
    public static void initialize() {
        Registry.register(Registries.ITEM_GROUP, CUSTOM_ITEM_GROUP_KEY, CUSTOM_ITEM_GROUP);

        ItemGroupEvents.modifyEntriesEvent(CUSTOM_ITEM_GROUP_KEY)
                .register((itemGroup) -> {
                    itemGroup.add(ModItems.TEST);
                    itemGroup.add(Items.ELYTRA);
                });
    }
}

File: src/main/java/ca/rafemurray/item/ModItems.kt
package ca.rafemurray.item

import ca.rafemurray.ElytraRacing
import ca.rafemurray.block.ModBlocks
import net.fabricmc.fabric.api.itemgroup.v1.FabricItemGroup
import net.minecraft.core.Registry
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.core.registries.Registries
import net.minecraft.network.chat.Component
import net.minecraft.resources.Identifier
import net.minecraft.resources.ResourceKey
import net.minecraft.world.item.CreativeModeTab
import net.minecraft.world.item.CreativeModeTab.ItemDisplayParameters
import net.minecraft.world.item.Item
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.Items
import java.util.function.Function

object ModItems {
    // Make and register a custom item group for the mod (creative inventory tab)
    private fun <T : Item?> register(name: String, itemFactory: Function<Item.Properties, T>, settings: Item.Properties): T {
        // Create the item key.
        val itemKey = ResourceKey.create(
            Registries.ITEM, Identifier.fromNamespaceAndPath(
                ElytraRacing.MOD_ID, name
            )
        )

        // Create the item instance.
        val item = itemFactory.apply(settings.setId(itemKey))

        // Register the item.
        Registry.register(BuiltInRegistries.ITEM, itemKey, item)

        return item
    }

    private val CUSTOM_CREATIVE_TAB_KEY: ResourceKey<CreativeModeTab> = ResourceKey.create(
        BuiltInRegistries.CREATIVE_MODE_TAB.key(), Identifier.fromNamespaceAndPath(ElytraRacing.MOD_ID, "creative_tab")
    )
    private val CUSTOM_CREATIVE_TAB: CreativeModeTab = FabricItemGroup.builder()
        .icon { ItemStack(Items.ELYTRA) }
        .title(Component.translatable("itemGroup.elytra-racing"))
        .displayItems { params: ItemDisplayParameters?, output: CreativeModeTab.Output ->
            output.accept(Items.ELYTRA)
            output.accept(COURSE_MARKER)
            output.accept(ModBlocks.BOOSTER)
            output.accept(ModBlocks.START)
            output.accept(ModBlocks.FINISH)
        }
        .build()

    private val COURSE_MARKER: Item = register(
        "course_marker",
        { settings: Item.Properties -> CourseMarkerItem(settings) }, Item.Properties()
    )

    fun initialize() {
        Registry.register(BuiltInRegistries.CREATIVE_MODE_TAB, CUSTOM_CREATIVE_TAB_KEY, CUSTOM_CREATIVE_TAB)
    }
}
